### PR TITLE
Download on fly

### LINKED ISSUE #174: Create data request zip files on the fly
We currently store zip files on the server, but that might cause problems with disk space. It might be better to create the zip files on the fly and not store them.

### PR DESCRIPTION
**Pull request by Eb-Zeero**
This change is [Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)

### CODE REVIEW

File: src/util/zipDataRequest.ts [hunk 1]
```diff
@@ -335,20 +241,59 @@ Pickering, T.E., Potter, S., Romero Colmenero, E., Vaisanen, P., Williams, T.,
...
+  based on the instrument and calibration level.
+ */
+  const dataFiles: any[] = [];
+  for (const df of artifacts) {
+    for (const calibrationLevel of Array.from(calibrationLevels)) {
+      let filepath: string;
+      let description: string;
+
+      if (calibrationLevel === "RAW") {
+        filepath = process.env.DATA_REQUEST_BASE_DIR + "/" + df.raw;
```
Comment (hettlage) [inline src/util/zipDataRequest.ts]: Must be the Fits base dir

**Review by Eb-Zeero — COMMENTED** [review]
> Reviewed 1 of 4 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: 5 of 8 files reviewed, 1 unresolved discussion (waiting on @hettlage)

___
*[src/util/zipDataRequest.ts, line 265 at r1](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MFpEnO6ZERkjnJ1A5UT-r1-265:-MG94QJ921tPR6THM4f6:bwzxbz5) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/5189bd3bba77db4ce7c03f0a2b055e6cc9f15433/src/util/zipDataRequest.ts#L265)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

Must be the Fits base dir
</blockquote></details>

ouch my bad done

**Review by hettlage — CHANGES_REQUESTED** [review]
> Reviewed 3 of 7 files at r1, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 5 unresolved discussions (waiting on @Eb-Zeero and @hettlage)

___
*[tslint.json, line 4 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmK8Je8-W5UuZeAZKs:-MGmK8Je8-W5UuZeAZKt:b1ltvzt) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/tslint.json#L4)):*
> ```jsonc
>   "extends": ["tslint:latest", "tslint-config-prettier"],
>   "rules": {
>     "no-console": false,
> ```

Shouldn't this be true?
___
*[src/createServer.ts, line 18 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmJjeW8oz2Xw9oacn2:-MGmJjeW8oz2Xw9oacn3:b-7pm5u8) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/createServer.ts#L18)):*
> ```XML
>   userDataLoader
> } from "./util/loaders";
> import { getUserById, mayViewDataFile, User } from "./util/user";
> ```

User might be an unnecessary import.
___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MGmKLP596gj0MExiaYG:buk1lj1) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
> ```GraphQL
>     """
>     requestedCalibrationLevels: [CalibrationLevel!]!
>   ): CreateDataRequestStatus!
> ```

Wouldn't it make more sense to return the id of the created data request? I dob;t think there is a need for status or message.
___
*[src/resolvers/dataRequest.ts, line 116 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmJXNk5iaPGjswwLla:-MGmJXNk5iaPGjswwLlb:bw1f7sm) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/resolvers/dataRequest.ts#L116)):*
> ```XML
>     return {
>       dataRequestId,
>       message: "The data request was successfully requested.",
> ```

requested --> made

But do we really need a message?

**Review by Eb-Zeero — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 5 unresolved discussions (waiting on @hettlage)

___
*[tslint.json, line 4 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmK8Je8-W5UuZeAZKs:-MGrBSyy2FjjQLrZv23Y:b-klgdo0) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/tslint.json#L4)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

Shouldn't this be true?
</blockquote></details>

this shoun't be here
___
*[src/createServer.ts, line 18 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmJjeW8oz2Xw9oacn2:-MGrC22bE_PnTHyjqiMZ:bbjp23x) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/createServer.ts#L18)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

User might be an unnecessary import.
</blockquote></details>

it is needed by this end point
`server.express.get("/?/data/:dataFileId/:dataFilename"...`
___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MGrCQOUA9DABlC_oF4S:bof5690) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

Wouldn't it make more sense to return the id of the created data request? I dob;t think there is a need for status or message.
</blockquote></details>

we do need the status. message doesn't matter.

it is easy to check if the status true that to check if an is exist.
___
*[src/resolvers/dataRequest.ts, line 116 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmJXNk5iaPGjswwLla:-MGrRBioBNqKqbEV377v:bkhihz6) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/resolvers/dataRequest.ts#L116)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

requested --> made

But do we really need a message?
</blockquote></details>

Removed message

**Review by hettlage — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 2 unresolved discussions (waiting on @Eb-Zeero and @hettlage)

___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MGrToDc1ZMuphauYVZt:b2py6q4) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
<details><summary><i>Previously, Eb-Zeero (Nhlavutelo Macebele) wrote…</i></summary><blockquote>

we do need the status. message doesn't matter.

it is easy to check if the status true that to check if an is exist.
</blockquote></details>

But will the status ever be false?

**Review by Eb-Zeero — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 2 unresolved discussions (waiting on @hettlage)

___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MHAzRNhDdqbh8uFOGaS:b-q8j6kr) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

But will the status ever be false?
</blockquote></details>

no there is an error

**Review by Eb-Zeero — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 2 unresolved discussions (waiting on @hettlage)

___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MHAzn079yehSTP8o30l:bnczh7c) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
<details><summary><i>Previously, Eb-Zeero (Nhlavutelo Macebele) wrote…</i></summary><blockquote>

no there is an error
</blockquote></details>

No it will never be
This mean when a user create a data request the only thing he should be accepting back is just an ID or an error

**Review by hettlage — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 1 unresolved discussion (waiting on @Eb-Zeero)

___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MHBDRoiAufzwO0_B9dt:b-31211f) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
<details><summary><i>Previously, Eb-Zeero (Nhlavutelo Macebele) wrote…</i></summary><blockquote>

No it will never be
This mean when a user create a data request the only thing he should be accepting back is just an ID or an error
</blockquote></details>

And you would not need a custom type. :-)

**Review by Eb-Zeero — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: 7 of 9 files reviewed, 1 unresolved discussion (waiting on @hettlage)

___
*[src/schema.graphql, line 198 at r5](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MGmKLP596gj0MExiaYF:-MHFa2kA8IhJ50rPrXDQ:b-896fix) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/ba7a98ed87ca3604a7b6f69e949b4a0672cfc7a6/src/schema.graphql#L198)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

And you would not need a custom type. :-)
</blockquote></details>

Done.

**Review by hettlage — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 3 unresolved discussions (waiting on @Eb-Zeero)

___
*[src/schema.graphql, line 358 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuR-7I8wgl_JZStVGY:-MHuR-7I8wgl_JZStVGZ:b7c0db0) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/schema.graphql#L358)):*
> ```GraphQL
>   user: User!
>   """
>   file location fro download.
> ```

fro ---> for
___
*[src/schema.graphql, line 398 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuQmGR3tS5WqpYsvsJ:-MHuQmGR3tS5WqpYsvsK:b-aj42ci) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/schema.graphql#L398)):*
> ```GraphQL
> 
> """
> The data request status.
> ```

I guess DataRequestStatus can be removed?
___
*[src/resolvers/dataRequest.ts, line 115 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuRPK55DabqP6MRTVM:-MHuRPK55DabqP6MRTVN:bllopic) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/resolvers/dataRequest.ts#L115)):*
> ```XML
>     await client.query("COMMIT");
> 
>     return {
> ```

You should return the uri field as well.

**Review by hettlage — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 3 unresolved discussions (waiting on @Eb-Zeero)

___
*[src/schema.graphql, line 358 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuR-7I8wgl_JZStVGY:-MHuR-7I8wgl_JZStVGZ:b7c0db0) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/schema.graphql#L358)):*
> ```GraphQL
>   user: User!
>   """
>   file location fro download.
> ```

fro ---> for
___
*[src/schema.graphql, line 398 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuQmGR3tS5WqpYsvsJ:-MHuQmGR3tS5WqpYsvsK:b-aj42ci) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/schema.graphql#L398)):*
> ```GraphQL
> 
> """
> The data request status.
> ```

I guess DataRequestStatus can be removed?
___
*[src/resolvers/dataRequest.ts, line 115 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuRPK55DabqP6MRTVM:-MHuRPK55DabqP6MRTVN:bllopic) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/resolvers/dataRequest.ts#L115)):*
> ```XML
>     await client.query("COMMIT");
> 
>     return {
> ```

You should return the uri field as well.

**Review by Eb-Zeero — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: 7 of 9 files reviewed, 2 unresolved discussions (waiting on @hettlage)

___
*[src/resolvers/dataRequest.ts, line 115 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MIrue-x59Jl3OGaWbTT:-MIrue-y8IWT_gzuc9ta:b-td8ker) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/resolvers/dataRequest.ts#L115)):*
> ```XML
>     await client.query("COMMIT");
> 
>     return {
> ```

Uri Field was contains data data request stored location which is no more it is no longer required

**Review by hettlage — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: all files reviewed, 4 unresolved discussions (waiting on @Eb-Zeero)

___
*[src/resolvers/dataRequest.ts, line 115 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MIrue-x59Jl3OGaWbTT:-MIsyuJ3128iXQ5ie1ar:bymlig3) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/resolvers/dataRequest.ts#L115)):*
<details><summary><i>Previously, Eb-Zeero (Nhlavutelo Macebele) wrote…</i></summary><blockquote>

Uri Field was contains data data request stored location which is no more it is no longer required
</blockquote></details>

But the front-end still needs to know where to find the file (even if it's created on the fly), and it would be natural to ask the server for this information.

**Review by Eb-Zeero — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/saltastroops/data-archive-backend/181)* status: 7 of 11 files reviewed, 3 unresolved discussions (waiting on @hettlage)

___
*[src/schema.graphql, line 358 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuR-7I8wgl_JZStVGY:-MIxI40sCxxXjf0af0Yi:b-896fix) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/schema.graphql#L358)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

fro ---> for
</blockquote></details>

Done.
___
*[src/schema.graphql, line 398 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuQmGR3tS5WqpYsvsJ:-MIxI3X_4ob-ZbzjWfjD:b-896fix) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/schema.graphql#L398)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

I guess DataRequestStatus can be removed?
</blockquote></details>

Done.
___
*[src/resolvers/dataRequest.ts, line 115 at r7](https://reviewable.io/reviews/saltastroops/data-archive-backend/181#-MHuRPK55DabqP6MRTVM:-MIxI1qd1S704HvobgV9:b-896fix) ([raw file](https://github.com/saltastroops/data-archive-backend/blob/4e5f001a4e43544616bfcc4d769eb9ec55a49cdb/src/resolvers/dataRequest.ts#L115)):*
<details><summary><i>Previously, hettlage wrote…</i></summary><blockquote>

You should return the uri field as well.
</blockquote></details>

Done.

**Pull request closed without merging**